include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,653 @@
+     *   ArrayView object from it; this is satisfied by the `std::vector` class,
+     *   among others.
+     *
+     * @dealiiRequiresUpdateFlags{update_gradients}
+     */
+    template <class InputVector>
+    void
+    get_function_gradients_from_local_dof_values(
+      const InputVector &dof_values,
+      std::vector<solution_gradient_type<typename InputVector::value_type>>
+        &gradients) const;
+
+  private:
+    /**
+     * The data structure that stores the renumbering of the degrees of freedom
+     * and of the quadrature points.
+     */
+    const RenumberingData &data;
+
+    /**
+     * Produce an inner vector compatible with the inner view, after copying the
+     * values with the correct numbering from the outer vector.
+     */
+    template <typename InputVector>
+    const InputVector &
+    outer_to_inner_dofs(const InputVector &outer_vector) const;
+
+    /**
+     * Produce an inner vector compatible with the inner view, and zero out its
+     * entries if necessary.
+     */
+    template <typename ValueType>
+    std::vector<ValueType> &
+    outer_to_inner_values(std::vector<ValueType> &outer_values) const;
+
+    /**
+     * Return the outer argument renumbered according to the quadrature
+     * renumbering. The values in the inner values are copied to the right
+     * position in the outer vector.
+     */
+    template <typename ValueType>
+    void
+    inner_to_outer_values(const std::vector<ValueType> &inner_values,
+                          std::vector<ValueType>       &outer_values) const;
+
+    /**
+     * Store a reference to the underlying view.
+     */
+    const ViewType &view;
+  };
+} // namespace FEValuesViews
+
+
+#ifndef DOXYGEN
+
+
+/*------------------------ Inline functions: namespace FEValuesViews --------*/
+
+namespace FEValuesViews
+{
+  template <typename ViewType>
+  RenumberedView<ViewType>::RenumberedView(const ViewType        &view,
+                                           const RenumberingData &data)
+    : view(view)
+    , data(data)
+  {}
+
+
+
+  template <typename Number>
+  inline std::string
+  RenumberingData::get_unique_container_name(
+    const std::string &prefix,
+    const unsigned int size,
+    const Number      &exemplar_number) const
+  {
+    return prefix + "_" + Utilities::int_to_string(size) + "_" +
+           Utilities::type_to_string(exemplar_number);
+  }
+
+
+
+  template <typename ViewType>
+  typename RenumberedView<ViewType>::value_type
+  RenumberedView<ViewType>::value(const unsigned int shape_function,
+                                  const unsigned int q_point) const
+  {
+    AssertIndexRange(shape_function, data.n_dofs);
+    AssertIndexRange(q_point, data.n_quadrature_points);
+
+    const auto inner_shape_function = data.dof_renumbering.empty() ?
+                                        shape_function :
+                                        data.dof_renumbering[shape_function];
```

Comment:
It's a shame we're still stuck with C++17. This would be so much nicer to write by storing the vector as a `std::optional` that either holds the vector or nothing, and then using the monadic operations...